File: tic-tac-toe-client/src/types/game.ts
export type Player = "X" | "O";

export type GameMode = "TwoPlayer" | "Computer";

export type GameStatus = "InProgress" | "Won" | "Draw";

export interface Move {
    moveNumber: number;
    player: Player;
    row: number;
    column: number;
}

export interface Game {
    id: string;
    boardSize: number;
    board: (Player | null)[];
    currentPlayer: Player;
    gameMode: GameMode;
    gameStatus: GameStatus;
    winner: Player | null;
    winningCells: number[];
    moveHistory: Move[];
}

export interface Scoreboard {
    xWins: number;
    oWins: number;
    draws: number;
}

File: TicTacToe.Api/Models/Game.cs
namespace TicTacToe.Api.Models;

public class Game
{
    public Guid Id { get; private set; }
    public int BoardSize { get; set; }
    public Player?[] Board { get; set; }
    public Player CurrentPlayer { get; set; } = Player.X;
    public GameMode GameMode { get; set; }
    public GameStatus GameStatus { get; set; } = GameStatus.InProgress;
    public Player? Winner { get; set; }
    public List<int> WinningCells { get; set; } = [];
    public List<Move> MoveHistory { get; set; } = [];

    public Game(Guid id, int boardSize = 3)
    {
        Id = id;

        BoardSize = boardSize;

        Board = new Player?[boardSize * boardSize];
    }

    public Game(int boardSize = 3) : this(Guid.NewGuid(), boardSize)
    {
    }
}